src/main/java/me/tomsdevsn/hetznercloud/HetznerCloudAPI.java

```diff
@@ -689,6 +690,195 @@ public DatacentersResponse getDatacenter(String name) {
+     * @return an ActionsResponse with the executed actions
+     */
+    public ActionResponse getActionOfFirewall(long firewallId, long actionId) {
+        return get(
+                String.format("%s/firewalls/%s/actions/%s", API_URL, firewallId, actionId),
+                ActionResponse.class);
+    }
+
+    /**
+     * Applies one Firewall to multiple resources.
+     *
+     * @param id of the firewall you want to add to resources
+     * @param applicationTargets you want to add
+     * @return an ActionsResponse with the executed actions
+     * @throws RuntimeException if the removalTargets don't pass {@link FWApplicationTarget#isValidOrThrow()}
+     */
+    public ActionsResponse applyFirewallToResources(long id, List<FWApplicationTarget> applicationTargets)
+            throws RuntimeException {
+        for (var target: applicationTargets)
+            target.isValidOrThrow();
+
+        return post(
+                API_URL + "/firewalls/" + id + "/actions/apply_to_resources",
+                Map.of("apply_to", applicationTargets),
+                ActionsResponse.class);
+    }
+
+    /**
+     * Removes one Firewall from multiple resources.
+     *
+     * @param id of the firewall you want to remove resources from
+     * @param removalTargets you want to remove
+     * @return an ActionsResponse with the executed actions
+     * @throws RuntimeException if the removalTargets don't pass {@link FWApplicationTarget#isValidOrThrow()}
+     */
+    public ActionsResponse removeFirewallFromResources(long id, List<FWApplicationTarget> removalTargets)
+            throws RuntimeException {
+        for (var target: removalTargets)
+            target.isValidOrThrow();
+
+        return post(
+                API_URL + "/firewalls/" + id + "/actions/remove_from_resources",
+                Map.of("remove_from", removalTargets),
+                ActionsResponse.class);
+    }
+
+    /**
+     * Removes all rules of a Firewall.
+     *
+     * @param id the firewall you want to remove the rules from
+     * @return an ActionsResponse with the executed actions
+     * @see #setFirewallRules(long, List)
+     */
+    public ActionsResponse removeAllRulesFromFirewall(long id) {
+        return setFirewallRules(id, Collections.EMPTY_LIST);
+    }
+
+    /**
+     * Sets the rules of a Firewall. All existing rules will be overwritten.
+     * If the firewallRules are empty, all rules are deleted.
+     *
+     * @param id of the Firewall you want to set the Rules on.
+     * @param firewallRules you want to set.
+     * @return an ActionsResponse with the executed actions
+     * @throws IllegalArgumentException if the size of firewallRules is greater than 50 (size > 50)
+     */
+    public ActionsResponse setFirewallRules(long id, List<FirewallRule> firewallRules)
+            throws IllegalArgumentException {
+        if (firewallRules.size() > 50)
+            throw new IllegalArgumentException("You can't add more than 50 rules");
```

Comment:
The check should be done from API in my opinion (this could change in the future).